### PR TITLE
Refactor event bindings of viewport

### PR DESCRIPTION
**Pull request by igoroctaviano**
**Work in progress**

### CODE REVIEW

File: src/CornerstoneViewport/CornerstoneViewport.js [hunk 1]
```diff
@@ -142,13 +142,18 @@ class CornerstoneViewport extends Component {
...
-    // EVENTS
-    this._bindInternalEventListeners();
-    this._bindExternalEventListeners();
+    // ~~ EVENTS: CORNERSTONE
     this._handleOnElementEnabledEvent();
+    this._bindInternalCornerstoneEventListeners();
+    this._bindExternalEventListeners("cornerstone");
 
-    // Fire 'er up
+    // Enable
```
Comment (dannyrb) [inline src/CornerstoneViewport/CornerstoneViewport.js]: Haha, probably a better comment

File: src/CornerstoneViewport/CornerstoneViewport.js [hunk 2]
```diff
@@ -142,13 +142,18 @@ class CornerstoneViewport extends Component {
...
     cornerstone.enable(this.element, cornerstoneOptions);
+
+    // ~~ EVENTS: ELEMENT
+    this._bindInternalElementEventListeners();
+    this._bindExternalEventListeners("element");
```
Comment (dannyrb) [inline src/CornerstoneViewport/CornerstoneViewport.js]: We technically need to unbind/rebind these anytime there's a disable/enable event

File: src/CornerstoneViewport/CornerstoneViewport.js [hunk 3]
```diff
@@ -281,17 +279,20 @@ class CornerstoneViewport extends Component {
...
-      this.setState({ isOverlayVisible });
-
-    // Rebind event listeners when props change to handle functional component method
-    this._bindExternalEventListeners(true);
-    this._bindExternalEventListeners();
+    /* Rebind event listeners when props change to handle functional component method
+    this._bindExternalEventListeners("cornerstone", true);
+    this._bindExternalEventListeners("element", true);
+    this._bindExternalEventListeners("cornerstone");
+    this._bindExternalEventListeners("element"); */
```
Comment (dannyrb) [inline src/CornerstoneViewport/CornerstoneViewport.js]: I think the thought here is that, at time of binding, we use values in props inside the bound methods. If we don't unbind/rebind, we could have the wrong values.

File: src/CornerstoneViewport/CornerstoneViewport.js (hunk 2 above)
Comment (igoroctaviano) [reply src/CornerstoneViewport/CornerstoneViewport.js]: This is being done in will mount/unmount already, callbacks were not being triggered when binding in every state change (will update cycle)

Comment (dannyrb) [reply src/CornerstoneViewport/CornerstoneViewport.js]: That's a good point. I guess if someone is manually enabling/disabling the element inside this component, they've already broken the abstraction. No reason to guard/protect for it

File: src/CornerstoneViewport/CornerstoneViewport.js [hunk 4]
```diff
@@ -281,17 +278,14 @@ class CornerstoneViewport extends Component {
...
     if (Object.keys(updatedState).length > 0) {
       this.setState(updatedState);
     }
-
-    if (isOverlayVisible !== prevIsOverlayVisible)
-      this.setState({ isOverlayVisible });
-
-    // Rebind event listeners when props change to handle functional component method
-    this._bindExternalEventListeners(true);
-    this._bindExternalEventListeners();
```
Comment (dannyrb) [inline src/CornerstoneViewport/CornerstoneViewport.js]: Huh. I guess none of these use props when binding. I'm not sure why we did unbind/rebind

**Pull request merged**